#' Generate filtered Bismark coverage files
#'
#' @param path_bismark2bedGraph The absolute path to the `bismark2bedGraph` function file
#' @param path_cpg_file The absolute path to the 'CpG_context' file generated by Bismark for the analyzed sample. This file must be produced using the `bismark_methylation_extractor` function with the -`-comprehensive` option, as strand-specific methylation information is not required. The file should contain five columns reporting in the following order: sequencing read ID, methylation status, chromosomal position, genomic position, and methylation call of the CpG covered by the sequencing read. For more details, consult the Bismark manual: https://felixkrueger.github.io/Bismark/bismark/.
#' @param path_read_table The Read Table for the analyzed sample, created using `METER::create_read_table` function, which summarizes methylation information for each sequencing read.
#' @param path_out The absolute path to the output folder. If not specified, the output folder will default to the same directory as the input 'CpG_context' file.
#' @param min_sites An integer specifying the minimum number of CpGs a read must contain to be included in the filtered "CpG_context" file and the corresponding filtered "coverage" file (default: min_sites = 6).
#' @param remove_cpg A logical value; set to TRUE to delete the intermediate 'CpG_context' file after the final 'coverage' file has been generated.
#'
#' @return A "coverage" file generated using the `bismark2bedGraph` function from Bismark, created considering only reads with an alpha value of 100% and covering a number of CpG sites greater than or equal to `min_sites`.
#' @export
filter_cov_alpha100 <- function(path_bismark2bedGraph, path_cpg_file, path_read_table, path_out = NULL, min_sites = 6, remove_cpg=FALSE){

  assertthat::assert_that(file.exists(path_bismark2bedGraph),
                          msg = "the bismark2bedGraph file path does not exist")

  assertthat::assert_that(file.exists(path_cpg_file),
                          msg = "the CpG_context file path does not exist")

  assertthat::assert_that(file.exists(path_read_table),
                          msg = "the Read Table file path does not exist")


  if (is.null(path_out)) {
    path_out <- file.path(dirname(path = path_cpg_file), 'filter_cov_alpha100')
  }

  dir.create(path = path_out, showWarnings = F)

  read_table <- readRDS(path_read_table)
  cpg_file <- data.table::fread(path_cpg_file, data.table = F)

  assertthat::assert_that(all(c("seq_id", "chr", "min_cpg_pos", "max_cpg_pos", "n_meth", "n_unmeth", "n_sites", "meth_perc", "dmr_id", "dmr_type") %in% colnames(read_table)),
                          msg = "The read_table does not include required columns")

  assertthat::assert_that(length(colnames(cpg_file))==5, msg = "The CpG file is not in the right format")


  ### check input
  read_table <- as.data.frame(read_table)
  cpg_file <- as.data.frame(cpg_file)


  ### FILTER READS
  tokeep <- read_table$seq_id[which(read_table$n_sites>=min_sites &
                                   (read_table$meth_perc==1 | read_table$meth_perc==0))]

  rm(read_table)

  ### FILTER cpg file
  cn_1 <- colnames(cpg_file)[1]

  ## create tmp colnames
  colnames(cpg_file) <- paste0('tmp_', 1:ncol(cpg_file))

  cpg_file <- cpg_file[which(cpg_file$tmp_1 %in% tokeep), ]
  rownames(cpg_file) <- NULL

  ## restore original colnames
  colnames(cpg_file) = c(cn_1, '', '', '', '')

  ### save filtered cpg file
  data.table::fwrite(cpg_file,
                     file = file.path(path_out, paste0('CpG_context_', gsub('.rds', '.txt.gz', basename(path_read_table), fixed = T))),
                     sep = '\t', compress = 'auto'
                     )


  ### run bismark bismark2bedGraph to create new cov file from the filtered CpG file
  in_file=file.path(path_out, paste0('CpG_context_', gsub('.rds', '.txt.gz', basename(path_read_table), fixed = T)))
  id=gsub('.rds', '', basename(path_read_table))

  system(paste(path_bismark2bedGraph, in_file, '--dir', path_out, '--output', id))

  if (remove_cpg==TRUE) {
    system(paste0('find ', path_out, ' -mindepth 1 -maxdepth 1 -type f -name "CpG_context*\\.txt\\.gz" -delete'))
    }

}



